Correction of Cumulative Green Roof Output Variables

RKStrand opened the pull request:
Pull request overview
---------------------
 - Fixes #9282 
 - It was noted that the cumulative output variable for precipitation were inconsistent with the sum of the precipitation output for the current value.  The sum of the current value should have been equal to or at the least very close to the cumulative value.  Unfortunately, that was not the case and there were three other output variables that exhibited the same trend.  This was identified as a problem with the way the output variables were set-up.  This has been fixed and now the output is consistent.  Differences were seen in the two input files that use the green roof and the output now looks correct/consistent.  As part of this work, the initializations for this module were separated out from the main routine and the initialization code was modernized for c++.

**NOTE: ENHANCEMENTS MUST FOLLOW A SUBMISSION PROCESS INCLUDING A FEATURE PROPOSAL AND DESIGN DOCUMENT PRIOR TO SUBMITTING CODE**

### Pull Request Author
Add to this list or remove from it as applicable.  This is a simple templated set of guidelines.
 - [X] Title of PR should be user-synopsis style (clearly understandable in a standalone changelog context)
 - [X] Label the PR with at least one of: Defect, Refactoring, NewFeature, Performance, and/or DoNoPublish
 - [X] Pull requests that impact EnergyPlus code must also include unit tests to cover enhancement or defect repair
 - [ ] Author should provide a "walkthrough" of relevant code changes using a GitHub code review comment process
 - [ ] If any diffs are expected, author must demonstrate they are justified using plots and descriptions
 - [ ] If changes fix a defect, the fix should be demonstrated in plots and descriptions
 - [ ] If any defect files are updated to a more recent version, upload new versions here or on DevSupport
 - [ ] If IDD requires transition, transition source, rules, ExpandObjects, and IDFs must be updated, and add IDDChange label
 - [ ] If structural output changes, add to output rules file and add OutputChange label
 - [ ] If adding/removing any LaTeX docs or figures, update that document's CMakeLists file dependencies

### Reviewer
This will not be exhaustively relevant to every PR.
 - [ ] Perform a Code Review on GitHub
 - [ ] If branch is behind develop, merge develop and build locally to check for side effects of the merge
 - [ ] If defect, verify by running develop branch and reproducing defect, then running PR and reproducing fix
 - [ ] If feature, test running new feature, try creative ways to break it
 - [ ] CI status: all green or justified
 - [ ] Check that performance is not impacted (CI Linux results include performance check)
 - [ ] Run Unit Test(s) locally
 - [ ] Check any new function arguments for performance impacts
 - [ ] Verify IDF naming conventions and styles, memos and notes and defaults
 - [ ] If new idf included, locally check the err file and other outputs

RKStrand committed:
Fix of Cumulative EcoRoof Issue

Cumulative EcoRoof output variables were not reporting properly because the variables were not set-up properly when the SetupOutputVariable calls were made.  This corrects that problem.

RKStrand committed:
Code simplification

Changes to the way the code is laid out to make it easier to read (more modern) and broken up into smaller chunks.  This should be equivalent to the previous commit as far as the algorithm, just better organized.

RKStrand committed:
Green Roof Additional Unit Tests

Creation of two unit tests to cover new subroutines that were created as part of the fix for some cumulative output variables not producing the correct values.

RKStrand committed:
Clang Format EcoRoofManager.hh

Forgot to clang format EcoRoofManager.hh.  Oops.  Sorry.

rraustad commented:
@RKStrand is this ready for review? This appears to correct a deficiency in reporting.

   < It was noted that the cumulative output variable for precipitation were inconsistent with the sum of the precipitation output

It would be helpful if a graph of before and after was presented for review and comparison.

RKStrand commented:
@rraustad Yes, this is ready for review.  Here are a couple of graphs for a design day for one of the four variables:

<img width="473" alt="image" src="https://user-images.githubusercontent.com/5804989/189318685-fd7e15cc-b840-4218-a2b5-edb260a3042d.png">
<img width="474" alt="image" src="https://user-images.githubusercontent.com/5804989/189318749-d0bae00a-aa53-4f00-b484-0ec4853d4b0e.png">
In the graphs, Debug Cumulative is the value reported by the debug branch for precipitation depth cumulative.  Fixed Cumulative is the same variable reported after the fix was implemented.  Current is the current value of precipitation depth.  Overall, when you sum up all of the current values you should get the cumulative number or very close to it since there will be a little lag due to time steps.  The Manual Cumulative number is manually summing up precipitation depth.

In the first graph, you can see that the debug version for cumulative is way off because the output variable was not properly defined.  In the fixed version for cumulative, it correctly tracks the cumulative value.  Again, there is a slight delay when compared to the "manual cumulative" because of time steps but the fixed and manual cumulative end up in the same place.  The second graph is focused in so you can see the current, fixed, and manual values better.  This was the variable that initially caused the defect to be logged.  I discovered during the fixed that the other three variables that I corrected also had this same problem.  For an annual run, all four cumulative variables in the debug version were no where close to the sum of all of the current values.  In the fixed version, the sum of each current variable for the entire now matches the cumulative value at the end of the year like it should.  I am only showing a snapshot here, but I can produce more data.  Is there more information that you would like to see?  If so, what else can I provide to help you as you review this?

Myoldmopar commented on src/EnergyPlus/EcoRoofManager.cc:
```diff
@@ -757,6 +540,204 @@ namespace EcoRoofManager {
+    void initEcoRoof(EnergyPlusData &state, int const SurfNum, int const ConstrNum)
+    {
+        // Using/Aliasing
+        using namespace DataEnvironment;
```
With this small function it's not important, but I think it's always better to remove usings and just be explicit at the usage.

Myoldmopar commented on src/EnergyPlus/EcoRoofManager.cc:
```diff
@@ -757,6 +540,204 @@ namespace EcoRoofManager {
+        using namespace DataEnvironment;
+
+        auto &thisMat = state.dataMaterial->Material(state.dataConstruction->Construct(ConstrNum).LayerPoint(1));
+        auto &thisEcoRoof = state.dataEcoRoofMgr;
```
I was going to say this is a weird `using`, but I think the weird part is that it appears we can only have one eco roof in a simulation file.  We can have multiple buildings, but only one eco roof.  Seems problematic, but nothing to worry about here.

Myoldmopar commented on src/EnergyPlus/EcoRoofManager.cc:
```diff
@@ -757,6 +540,204 @@ namespace EcoRoofManager {
+                            thisEcoRoof->CumRunoff,
+                            OutputProcessor::SOVTimeStepType::Zone,
+                            OutputProcessor::SOVStoreType::State,
+                            "Environment");
```
With the refactor, it's hard to tell exactly what change was made.  However, based on the original issue I assumed it was a change to the setup output variable calls, and it does appear the arguments are different now, and this seems to be good.

Myoldmopar reviewed (approved):
Looks good to me.

Myoldmopar commented:
Diffs in the eco roof files as expected, with math diffs reflecting the numeric change, and rdd changes reflecting the variable types.  Building locally.

Myoldmopar commented:
Everything is happy locally, merging this one in.  Thanks @RKStrand
